cli/cmd/compose/run.go

```diff
@@ -0,0 +1,111 @@
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package compose
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/compose-spec/compose-go/cli"
+	"github.com/spf13/cobra"
+
+	"github.com/docker/compose-cli/api/client"
+	"github.com/docker/compose-cli/api/compose"
+	apicontext "github.com/docker/compose-cli/context"
+	"github.com/docker/compose-cli/context/store"
+	"github.com/docker/compose-cli/progress"
+)
+
+type runOptions struct {
+	composeOptions
+
+	Name        string
+	Command     []string
+	WorkingDir  string
+	Environment []string
+	Detach      bool
+	Publish     []string
+	Labels      []string
+	Volumes     []string
+	Rm          bool
+}
+
+func runCommand() *cobra.Command {
+	opts := runOptions{}
+	runCmd := &cobra.Command{
+		Use:   "run [options] [-v VOLUME...] [-p PORT...] [-e KEY=VAL...] [-l KEY=VALUE...] SERVICE [COMMAND] [ARGS...]",
+		Short: "Run a one-off command on a service.",
+		Args:  cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			s := store.ContextStore(cmd.Context())
+			currentCtx, err := s.Get(apicontext.CurrentContext(cmd.Context()))
+			if err != nil {
+				return err
+			}
+			switch currentCtx.Type() {
+			case store.LocalContextType:
+			default:
+				return fmt.Errorf(`Command "run" is not yet implemented for %q context type`, currentCtx.Type())
+			}
+			if len(args) > 1 {
+				opts.Command = args[1:]
+			}
+			opts.Name = args[0]
+			return runRun(cmd.Context(), opts)
+		},
+	}
+	runCmd.Flags().StringVar(&opts.WorkingDir, "workdir", "", "Work dir")
+
+	runCmd.Flags().SetInterspersed(false)
+
+	runCmd.Flags().StringArrayVarP(&opts.Publish, "publish", "p", []string{}, "Publish a container's port(s). [HOST_PORT:]CONTAINER_PORT")
+	runCmd.Flags().StringVar(&opts.Name, "name", "", "Assign a name to the container")
+	runCmd.Flags().StringArrayVarP(&opts.Labels, "label", "l", []string{}, "Set meta data on a container")
+	runCmd.Flags().StringArrayVarP(&opts.Volumes, "volume", "v", []string{}, "Volume. Ex: storageaccount/my_share[:/absolute/path/to/target][:ro]")
+	runCmd.Flags().BoolVarP(&opts.Detach, "detach", "d", false, "Run container in background and print container ID")
+	runCmd.Flags().StringArrayVarP(&opts.Environment, "env", "e", []string{}, "Set environment variables")
+	runCmd.Flags().BoolVar(&opts.Rm, "rm", false, "Automatically remove the container when it exits")
+
+	//addComposeCommonFlags(runCmd.Flags(), &opts)
+	return runCmd
+}
+
+func runRun(ctx context.Context, opts runOptions) error {
+
+	c, err := client.New(ctx)
+	if err != nil {
+		return err
+	}
+
+	_, err = progress.Run(ctx, func(ctx context.Context) (string, error) {
```

Comment:
`compose run` has to attach stdin/stdout to the service container, not compatible with running inside a progress context.
see UpCommand's `CreateAndStart`

Reply:
see https://github.com/docker/compose-cli/pull/1015/files#diff-338f90b88c22fe2a1befe8f111f6fc425f7e7d4f0ee375d23d925d5246a98d7bR405
will need to be exposed on compose/api so that you can implement `run` as `Create+Attach+Start`

Reply:
Should we expose the attach method, or have something like to the current up flow, but add an API `compose.Start(ctx, project, servicename, consumer)`, similar to compose.Up(ctx, project, consumer)` ?

Reply:
Imho it would make sense to have a `compose.Attach()` method in the API, as here we don't just want to collect stdout, but also want to bind stdin.

Reply:
`docker-compose run` is interactive by default